app/main.py

```diff
@@ -76,3 +105,181 @@ def get_share_link():
+        abort(NOT_FOUND, description=str(e))
+
+
+def is_json(json_data):
+    try:
+        json.loads(json_data)
+    except ValueError:
+        return False
+    return True
+
+
+@app.route("/nodes/<node_type>", methods=["POST"])
+# Exports all records in a dictionary node
+def export_node(node_type):
+    post_data = request.get_json()
+    program = post_data.get("program")
+    project = post_data.get("project")
+    format = post_data.get("format")
+    if program == None or project == None or format == None:
+        abort(BAD_REQUEST)
+
+    res = requests.get(
+        f"{Gen3Config.GEN3_ENDPOINT_URL}/api/v0/submission/{program}/{project}/export/?node_label={node_type}&format={format}", headers=HEADER)
+
+    json_data = json.loads(res.content)
+    if is_json(res.content) and "data" in json_data and json_data["data"] != []:
+        return res.content
+    else:
+        abort(NOT_FOUND)
+
+
+@app.route("/records/<uuids>", methods=["POST"])
+# Exports one or more records(records must in one node), use comma to separate the uuids
+# e.g. uuid1,uuid2,uuid3
+def export_record(uuids):
+    post_data = request.get_json()
+    program = post_data.get("program")
+    project = post_data.get("project")
+    format = post_data.get("format")
+    if program == None or project == None or format == None:
+        abort(BAD_REQUEST)
+
+    res = requests.get(
+        f"{Gen3Config.GEN3_ENDPOINT_URL}/api/v0/submission/{program}/{project}/export/?ids={uuids}&format={format}", headers=HEADER)
+
+    json_data = json.loads(res.content)
+    if b"id" in res.content:
+        return res.content
+    else:
+        abort(NOT_FOUND, description=json_data["message"])
+
+
+@app.route("/graphql", methods=["POST"])
+# Only used for filtering the files in a specific node for now
+def graphql_filter():
+    post_data = request.get_json()
+    node_type = post_data.get("node_type")
+    # Condition post format should looks like
+    # '(project_id: ["demo1-jenkins", ...], tissue_type: ["Contrived", "Normal", ...], ...)'
+    condition = post_data.get("condition")
+    # Field post format should looks like
+    # "submitter_id tissue_type tumor_code ..."
+    field = post_data.get("field")
+    if node_type == None or condition == None or field == None:
+        abort(BAD_REQUEST)
+
+    query = {
+        "query":
+        """{""" +
+        f"""{node_type}{condition}""" +
+        """{""" +
+        f"""{field}""" +
+        """}""" +
+        """}"""
+    }
+    res = requests.post(
+        f"{Gen3Config.GEN3_ENDPOINT_URL}/api/v0/submission/graphql/", json=query, headers=HEADER)
+
+    json_data = json.loads(res.content)
+    if json_data["data"] == None:
+        abort(NOT_FOUND, description=json_data["errors"])
+    else:
+        if json_data["data"][f"{node_type}"] == []:
+            abort(NOT_FOUND)
+        else:
+            return res.content
+
+
+@app.route("/<program>/<project>/<uuid>/<format>/download", methods=["GET"])
```

Comment:
One last thing I noticed, I would prefer specifying the parameters after the route, i.e /download/<program>/<project>/<uuid>/<format>
Thanks

Reply:
Yes, I have modified it.